### PR TITLE
Add initial technical interviewing workshop

### LINKED ISSUE #175: [Workshop] Technical Interview Preparation
# Workshop Request

## What is the title of your workshop

Technical Interview Preparation

## What are the key topics that will be covered in the workshop?

* Typical structures of technical interviews
* How to approach problems
* Important things to remember/focus on (e.g. asking clarifying questions, explaining what you're thinking, explaining what you're stuck on, clear naming, ...)
* Handling changing constraints, or looking for improvements to your code (a lot of technical interviews take a form "Solve a problem", "Now here's an extra constraint, how does your code handle that?", which is a skill we haven't really given trainees much experience with)
* Worked example question(s), maybe actually asking some in small groups

As a stretch goal, we _may_ want to try to touch on some hand-waving efficiency (e.g. showing two solutions to a problem, one of which memoises to avoid an array traversal inside a loop, or something), and discussing how one solution is "faster" or "uses less memory" than another, but definitely not going as far as trying to formalise this (i.e. no big-O).

## What knowledge does the student need before starting?

Given a description of a set of inputs/behaviours/outputs, to be able to write a function to implement the behaviour. Example problems in terms of scope/complexity are:
* Find the longest word in an array of strings, where each string may contain arbitrarily many words
* Compute the longest prefix common to {all,any two} strings in an array
* Parse a string of Roman numerals to decimal
* Write an object which has a method to tell it the winner and loser of a game, and returns a ranking of all known players

In particular, this requires knowledge of:
* All of JS1 (functions, strings, arrays, general datatypes, comparisons, control flow)
* Keeping and mutating state in objects

## What are the topics that will **not** be covered in the workshop?

* Any coding techniques. This workshop covers approaches, and context, but assumes pre-existing knowledge of all relevant techniques in any programming language (and will be delivered assuming JavaScript is the language known).
* "System design" style interviews - though some of the techniques cross-apply, open-ended systems design requires a lot more context than solving relatively narrowly defined coding problems.

## Any other notes?

Some discussion happened on https://github.com/CodeYourFuture/syllabus/issues/158 around this, from which this issue spun out. In particular, we should work out how much of methods/object state we want to teach in JS2, and if that amount is little/none, we should work out where we _should_ teach this content.

We should probably perform a survey of what knowledge is required to successfully interview at different companies, to make sure we're covering roughly the range we're targeting.

## Completion

When completed the lesson plan, slides and exercises should be added to the Workshop section of the Syllabus

### PR DESCRIPTION
**Pull request by illicitonion**
I imagine this being about 60-90 minutes of explaining, and then
splitting off to do lots of hands-on practice with TAs.

There are a few TODOs, and we should flesh out some instructor notes,
but I wanted to get some early feedback in :)

Fixes #175

-----
[View rendered docs/workshops/interviews/index.md](https://github.com/CodeYourFuture/syllabus/blob/technical-interview-workshop/docs/workshops/interviews/index.md)
[View rendered docs/workshops/interviews/technical/sample-questions.md](https://github.com/CodeYourFuture/syllabus/blob/technical-interview-workshop/docs/workshops/interviews/technical/sample-questions.md)
[View rendered docs/workshops/interviews/technical/workshop.md](https://github.com/CodeYourFuture/syllabus/blob/technical-interview-workshop/docs/workshops/interviews/technical/workshop.md)

### CODE REVIEW

**Comment by SallyMcGrath** [comment]
This is a great start. What do you think about structuring the workshop as a series of interviews? Have (at least) two teachers staging the explanation, more like this:

Interviewer: What is your goal in a technical interview?
Interviewee: My goal in a technical interview is to show that I'm going to be someone you want to work with.
Interviewer:  How will you show that? 
Interviewee: I'll definitely be friendly - no one likes someone getting angry with them! I'll try to be patient.  [...]

And provide the meta narrative as notes covering the same thing.

File: docs/workshops/interviews/technical/workshop.md [hunk 1]
```diff
@@ -0,0 +1,262 @@
...
+
+> I need to find the split the string into words, and then I need to find the logest word...
+>
+> To split the string into words, I need to remove any punctuation, and then split it on whitespace.
+>
+> Let's start with that, and I can work out the longest word part after.
+>
+> Ok, so I know I can use String.replace to get rid of the punctuation, as long as I know what characters are punctuation... But I'm not sure how to work out which characters are punctuation.
+
+At which point your interviewer may say something like "Let's assume there's no punctuation" (because they're not interested in that part), or "assume you can call a function called stripPunctuation()" (because they know we use libraries for lots of things, and this isn't what they want to focucs on), or they may give you a hint, like "could we use a regular expression to find all of the characters that _aren't_ letters or whitespace?". But **if the interviewer doesn't know what you're thinking, they can't help** - and they want to help.
```
Comment (alismith-dev) [inline docs/workshops/interviews/technical/workshop.md:150]: I'm not sure this is the best place to put this, but we might want to cover something about googling. My sense is that most interviews these days are open book, so looking up specifics of regex would be ok.

It might also be worth stating that googling "find the longest word in a sentence in javascript" is probably not appropriate 😉

File: docs/workshops/interviews/technical/workshop.md [hunk 2]
```diff
@@ -0,0 +1,262 @@
...
+
+* Friendly - no one likes someone getting angry with them!
+* Patient.
+* Thoughtful - jumping straight into your first idea doesn't make it likely that you're doing the best thing.
+* Open to feedback - if you have an idea that may help me, it's frustrating if I don't listen to it.
+* Communicative - if I'm stuck, and you don't know what I'm thinking, it's hard for you to participate.
+* Able to think through a problem.
+* Able to write code that works.
+
+Notice that **most of these are not technical traits**. While you're solving technical problems in a technical interview, most of the point of the interview is to see how you interact with people, and how you think.
```
Comment (alismith-dev) [inline docs/workshops/interviews/technical/workshop.md:37]: I wonder if there's some value in giving insight into the interviewer's point of view here, to reinforce the point.

We could note that most interviewers will have a checklist similar to the list above that they have to say "I saw evidence of X" / "I didn't see evidence of X", and therefore it's your job to make sure they can check things off the list.

Comment (alismith-dev) [reply docs/workshops/interviews/technical/workshop.md:37]: This comes up as a theme throughout, so an alternative suggestion would be to recap all the things mentioned at the end as a summary.

File: docs/workshops/interviews/technical/workshop.md [hunk 3]
```diff
@@ -0,0 +1,262 @@
...
+## Answering questions about your solution
+
+Your interviewer may now ask you some questions about your solution.
+
+Some examples of things the interviewer may ask about:
+* Why you chose to do something one way rather than another (e.g. why you used a `for` loop instead of a call to `map`).
+* Whether there are any interesting edge cases your code doesn't handle.
+* If there is anything you would change about the code.
+* TODO: Think about how to discuss: If there are any trade-offs you've made in your code
+* TODO: Think about whether to discuss efficiency at all.
```
Comment (alismith-dev) [inline docs/workshops/interviews/technical/workshop.md:243]: I would be in favour of discussing this - and making a note here that we might want to mention that's also referred to as performance - although I would probably be against getting into Big-O given the time constraints.

File: docs/workshops/interviews/technical/workshop.md [hunk 4]
```diff
@@ -0,0 +1,262 @@
...
+* TODO: Expand this list.
+
+:::tip
+Asking you about other ways of solving the problem does not mean you did it wrong! It can still be interesting to talk about other ways to do things, even if yours is a good one!
```
Comment (alismith-dev) [inline docs/workshops/interviews/technical/workshop.md:247]: I wonder if it's worth mentioning that the interviewer has probably seen lots of solutions to the problem, and therefore usually has a much better understanding than you do when seeing it for the first time.

File: docs/workshops/interviews/technical/workshop.md [hunk 5]
```diff
@@ -0,0 +1,262 @@
...
+:::
+
+## Handling changing constraints
+
+After you've finished solving the problem, your interviewer may ask you to change your code, or change the problem a bit. **This does not mean anything was wrong with your code**, it's a normal way to expand a problem.
```
Comment (alismith-dev) [inline docs/workshops/interviews/technical/workshop.md:252]: Sorry for the typesetting nitpick but through most of this document, we're using _italics_ for emphasis, but here we're switching to **bold**. Do you think it's worth making them consistent.

Fwiw, in the React module, I tended to use _italics_ for new terms (I was trying to avoid "scare quotes"), and **bold** for emphasis.

File: docs/workshops/interviews/technical/workshop.md [hunk 6]
```diff
@@ -0,0 +1,262 @@
...
+* The interviewer may change the problem a bit, and ask you to solve it (again, either by writing it down, or just by talking about it).
+
+You may be asked to solve one problem, or more than one, but ordinarily you'll spend somewhere between 15 and 60 minutes on each problem you're asked.
+
+The interviewer probably doesn't care what programming language you use (so you should probably use Javascript!), and will say if they do.
+
+TODO: Work out how to advise trainees get this information: _Some_ interviewers care a lot about details (like remembering every `;`, and that you remember whether a function is called `startsWith` or `beginsWith` or `hasPrefix`). Others don't. It's worth asking how precise you need to be.
+
+:::tip Remember
+The goal is not to get to a solution as fast as possible. The goal is to show that you're going to be a good person to work with.
```
Comment (alismith-dev) [inline docs/workshops/interviews/technical/workshop.md:58]: I might emphasise this even more by making it something like: Step 1: Take a deep breath and don't panic 🙂

File: docs/workshops/interviews/technical/workshop.md [hunk 7]
```diff
@@ -0,0 +1,262 @@
...
+:::
+
+Let's work through an example interview question to get a feel for how this works:
```
[the remark this answers is not in the document]
Comment (alismith-dev) [reply docs/workshops/interviews/technical/workshop.md:61]: Actually reviewing the whole thing, I don't think a whole separate workshop would be necessary, since most of the points still apply.

I think it could be handled with some notes at the end comparing the different styles.

**Review by alismith-dev — COMMENTED** [review]
Thanks for writing this up @illicitonion! I've left a few comments, but overall looks great.

My main bit of feedback is around different styles of interviews. However, I don't think we should change the structure of this as trying to cover everything would get confusing and I think most of the advice applies regardless of the style. I'd be in favour of a section which compares different styles and what the expectations might be.

File: docs/workshops/interviews/technical/workshop.md [hunk 8]
```diff
@@ -0,0 +1,262 @@
+---
+id: workshop
+title: Technical Interviews
+sidebar_label: Technical Interviews
+---
+
+# Technical Interviews
```
Comment (ChrisOwen101) [inline docs/workshops/interviews/technical/workshop.md:7]: Sorry to be this person but - could we get some Learning Objectives here? What exactly are we trying to achieve with this workshop?

File: docs/workshops/interviews/technical/workshop.md [hunk 9]
```diff
@@ -0,0 +1,262 @@
...
+
+For example, with our example question, the interviewer may tell you that, actually, words containing punctuation should be completely ignored and must never be returned, or that punctuation inside words counts as a letter. The interviewer may ask you to change your code, or may just ask you to talk through how you would change it without actually doing so.
+
+When making this kind of change, follow exactly the same steps as you did the first time - make sure you understand the problem, talk through any edge cases, and remember to keep thinking aloud!
+
+Follow-up questions and changes **do not mean that you made a mistake**. They're a normal part of an interview - don't worry about them.
+
+## Workshop
+
+We're now going to split into breakout rooms and practice some interview questions together. Your TA will have a list of technical interview questions to try out.
```
Comment (ChrisOwen101) [inline docs/workshops/interviews/technical/workshop.md:262]: Could we add rubric or guide for how these could be assessed? It'd be cool to give the students something at the end of this that says 

"Your speaking out loud was 4/5 but you need to work on asking for help."

It needn't be very heavy touch, just something that gets our TAs in the mindset that they should be accessing to some degree.

**Review by ChrisOwen101 — APPROVED** [review]
Really great work here! Thanks Daniel!

I wonder how useful prepping the students for Whiteboard interviews is - I don't expect they'll be doing them but I suppose it doesn't hurt. If I was to reduce scope in one place, that's where it'd be. 

How're you expecting the TAs to run the sessions? Collaborative code editor?

Pre-emptive approve after Ali's comments

File: docs/workshops/interviews/technical/sample-questions.md [hunk 10]
```diff
@@ -0,0 +1,39 @@
+---
+id: sample-questions
+title: Sample questions
+sidebar_label: Sample Questions
+---
+
+# Sample questions for technical interviews
```
Comment (ChrisOwen101) [inline docs/workshops/interviews/technical/sample-questions.md:7]: Could we add some instruction for TAs about how to run this? e.g. setup collavorative coding env,

**Comment by gregdyke** [comment]
Some random thoughts:

- If students are practiced in TDD/TDD as a problem solving strategy, I think we should encourage them to showcase it during the interview. They can clarify their understanding of the problem and their understanding of the edge cases by writing out some tests
- Some interviews aim to find out if you can solve problems with programming. Others aim to find out if you actually understand/know some of the details of X programming language (and that might be the difference between being happy with startsWith vs beginsWith?). Worth mentioning? (are the latter frequent?)

**Review by SallyMcGrath — APPROVED** [review]
